### PR TITLE
Parmest utils submodule, added support for Params and Vars

### LINKED ISSUE #2281: parmest enhancement: support params or variables
## Summary

As part of the IDAES end-to-end uncertainty discussion, it came to light that:
- Parameter estimation requires uncertainty model parameters to be defined as Pyomo variables
- Robust optimization (PyROS) requires uncertain model parameters to be defined as Pyomo parameters (mutable)

Recommendation/Feature Request:
- Generalize parmest to support either params or variables for uncertainty parameters that need to be regressed
- PyROS already has some code to handle this conversion under the hood
- Long-term: develop general utility code to support this conversion under the hood, we can use such a utility in parmest, PyROS, and the entire end-to-end uncertainty framework

## Rationale

The distinction between variables and parameters becomes confusing to users, especially new users.

### PR DESCRIPTION
**Pull request by kaklise**
## Summary/Motivation:
This PR starts a utils submodule in parmest which contains functions that could be used within other Pyomo contrib packages.  Currently, the submodule includes functionality from mpi-sppy (in `create_ef.py` and `scenario_tree.py`), an mpi task manager (in `mpi_utils.py`), and ipopt solver diagnostics (in `ipopy_solver_wrapper.py`).  These files were all moved directly from the main parmest folder into parmest/utils.  

An an additional utility function was added to convert model Params to Vars.  The function that converts Params to Vars builds on methods in PyROS.  Some modifications were made to retain the names from theta list (indexed parameters create indexed variables, non indexed parameters create non indexed variables).  The reactor design model was updated to define unknown parameters as Params.  Tests include cases that use Param, IndexedParam, Var, and IndexedVar.

## Changes proposed in this PR:
- Added parmest utils submodule
- Added ability to convert model Params into Vars

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by blnicho** [comment]
@kaklise try merging main into your branch to see if that resolves the pypy failure

File: pyomo/contrib/parmest/tests/test_utils.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
```
Comment (blnicho) [inline pyomo/contrib/parmest/tests/test_utils.py:9]: Copyright statement needs to be updated. Look at `parmest.py` for the updated one.

File: pyomo/contrib/parmest/utils/__init__.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
```
Comment (blnicho) [inline pyomo/contrib/parmest/utils/__init__.py:9]: Copyright statement needs to be updated. Look at `parmest.py` for the updated one.

File: pyomo/contrib/parmest/utils/model_utils.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
```
Comment (blnicho) [inline pyomo/contrib/parmest/utils/model_utils.py:9]: Copyright statement needs to be updated. Look at `parmest.py` for the updated one.

File: pyomo/contrib/parmest/utils/__init__.py [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+from .create_ef import *
+from .ipopt_solver_wrapper import *
+from .model_utils import *
+from .mpi_utils import *
+from .scenario_tree import *
```
[the remark this answers is not in the document]
Comment (blnicho) [reply pyomo/contrib/parmest/utils/__init__.py:14]: The consensus at the dev call was that we should include deprecation warnings for these moved imports. We should also not use "import *" syntax.

File: pyomo/contrib/parmest/parmest.py [hunk 5]
```diff
@@ -327,17 +326,30 @@ def _create_parmest_model(self, data):
...
+        
+        # Add objective function (optional)
+        if self.obj_function:
+            for obj in model.component_objects(pyo.Objective):
+                obj.deactivate()
+        
+            def FirstStageCost_rule(model):
+                return 0
+            model.FirstStageCost = pyo.Expression(rule=FirstStageCost_rule)
+            model.SecondStageCost = pyo.Expression(rule=_SecondStageCostExpr(self.obj_function, data))
```
[the remark this answers is not in the document]
Comment (kaklise) [reply pyomo/contrib/parmest/parmest.py]: Added a check to ensure FirstStageCost, SecondStageCost and Total_Cost_Objective don't already exist.

File: pyomo/contrib/parmest/parmest.py [hunk 6]
```diff
@@ -327,17 +326,30 @@ def _create_parmest_model(self, data):
...
             model.parmest_dummy_var = pyo.Var(initialize = 1.0)
+        
+        # Add objective function (optional)
+        if self.obj_function:
+            for obj in model.component_objects(pyo.Objective):
+                obj.deactivate()
+        
+            def FirstStageCost_rule(model):
+                return 0
+            model.FirstStageCost = pyo.Expression(rule=FirstStageCost_rule)
```
[the remark this answers is not in the document]
Comment (kaklise) [reply pyomo/contrib/parmest/parmest.py]: done

File: pyomo/contrib/parmest/utils/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (kaklise) [reply pyomo/contrib/parmest/utils/__init__.py]: Added full import statements and used `relocated_module_attribute` in the top level __init__ file to include deprecation warnings

**Review by blnicho — APPROVED** [review]

**Comment by blnicho** [comment]
@kaklise just FYI, we're seeing these same test failures everywhere and working on a fix.

**Pull request merged**